### DISCUSSION TITLE
SSD202D boot process

### DISCUSSION THREAD

**Discussion by Michel-FK** [linux-chenxing/linux-chenxing.org#45] [General]
Coming from an Allwinner background, I try to understand the SSD302D boot process.

Right now, it looks very confusing:

- the Boot ROM is first loading an Initial Program Loader (IPL) from either NAND/NOR SPI Flash into SRAM
- but is the Boot ROM able to load IPL directly from SD Card? I would like to boot directly from SD Card if possible, in order to avoid adding a chip on board, I am not sure this is even possible
- it looks like the IPL is then loading an IPL_CUST, but for what purpose? And from where?
- then, the U-Boot SPL (Secondary Program Loader) is loaded into SDRAM, but from where (SD card?), and what is its purpose if SDRAM is already initialized?
- U-Boot itself is loaded
- Eventually, the DTB and kernel image are loaded and the kernel is executed

**Comment by fifteenhex** [comment]
> ROM is first loading an Initial Program Loader (IPL) from either NAND/NOR SPI Flash into SRAM

Yes.

> * but is the Boot ROM able to load IPL directly from SD Card? I would like to boot directly from SD Card if possible, in order to avoid adding a chip on board, I am not sure this is even possible

On the SSD202D the options for the boot rom seem to be SPI NOR or SPI NAND

> * it looks like the IPL is then loading an IPL_CUST, but for what purpose? And from where?

IPL runs in SRAM (~48K). IPL_CUST runs in DRAM. You can do more in DRAM.
For the mainline-ish boot process IPL_CUST is replaced with the u-boot SPL.

> * then, the U-Boot SPL (Secondary Program Loader) is loaded into SDRAM, but from where (SD card?), and what is its purpose if SDRAM is already initialized?

SPL isn't used in the vendor boot flow. It's used in the mainline-ish bootflow.
You could potentially go from IPL to u-boot proper if you want but I use the SPL as a backup for when I break the main u-boot as the SPL can reload u-boot over serial.

> * U-Boot itself is loaded
> * Eventually, the DTB and kernel image are loaded and the kernel is executed

**Reply by Michel-FK** [reply]
Does anybody have a Boot ROM dump?
I would like to disassemble it in order to make sure what are the exact boot options.
I did it for the Allwinner V3s (https://github.com/FunKey-Project/Allwinner-V3s-BROM).

Otherwise I will have to wait until I receive my board.

**Reply by ddB0515** [reply]
If you wanted to play a bit with SDK you can get it from [here](http://doc.industio.com/docs/ssd20x_manual/ssd20x_manual-1cradqrvf230l)
Also would help as much as I can but never did such things so if I can do anything let me know :)

**Reply by Michel-FK** [reply]
Out of the 16KB Boot ROM, only 5KB (0x0000-0x136d) are effectively used, everything else is just zeros except for 32 bytes at the end, and no sign of any SD Card boot capability, so only SPI NAND/NOR :-(

It's a shame, as it is very easy to shoehorn a r/o FAT filesystem driver that would have been able to load the SPL into SRAM directly from SD-Card or eMMC...
